File: sdk/ruby/test/wire/webhooks_test.rb
# frozen_string_literal: true

require_relative "wiremock_test_case"

class WebhooksWireTest < WireMockTestCase
  def setup
    super

    @client = AgentTrunk::Client.new(
      access_token: "<token>",
      base_url: WIREMOCK_BASE_URL
    )
  end

  def test_webhooks_list_with_wiremock
    test_id = "webhooks.list.0"

    @client.webhooks.list(
      trunk_id: "trunkId",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "webhooks.list.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/trunks/trunkId/webhooks",
      query_params: nil,
      expected: 1
    )
  end

  def test_webhooks_create_portal_with_wiremock
    test_id = "webhooks.create_portal.0"

    @client.webhooks.create_portal(
      trunk_id: "trunkId",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "webhooks.create_portal.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/trunks/trunkId/webhooks/portal",
      query_params: nil,
      expected: 1
    )
  end

  def test_webhooks_retry__with_wiremock
    test_id = "webhooks.retry_.0"

    @client.webhooks.retry_(
      trunk_id: "trunkId",
      event_id: "eventId",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "webhooks.retry_.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/trunks/trunkId/webhooks/eventId/retry",
      query_params: nil,
      expected: 1
    )
  end
end

File: sdk/rust/tests/webhooks_test.rs
use agenttrunk::prelude::*;

mod wire_test_utils;

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_webhooks_list_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentTrunk::new(config).expect("Failed to build client");

    let result = client
        .webhooks
        .list(
            &"trunkId".to_string(),
            &WebhooksListQueryRequest {
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("GET", "/v1/trunks/trunkId/webhooks", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_webhooks_create_portal_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentTrunk::new(config).expect("Failed to build client");

    let result = client
        .webhooks
        .create_portal(&"trunkId".to_string(), None)
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("POST", "/v1/trunks/trunkId/webhooks/portal", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_webhooks_retry_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentTrunk::new(config).expect("Failed to build client");

    let result = client
        .webhooks
        .retry(&"trunkId".to_string(), &"eventId".to_string(), None)
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "POST",
        "/v1/trunks/trunkId/webhooks/eventId/retry",
        None,
        1,
    )
    .await
    .unwrap();
}
